java/pattern/schema/Rule.java

```diff
@@ -127,6 +129,11 @@ private static void validateThen(String label, @Nullable Conjunction<? extends P
             throw GraqlException.of(INVALID_RULE_THEN.message(label, then));
         }
 
+        // rule 'has' cannot assign both an attribute type and a variable
+        if (numConstraints == 1 && then.has().size()==1 && then.has().get(0).type().isNamed() && then.has().get(0).attribute().isVariable()) {
+            throw GraqlException.of(INVALID_RULE_THEN_HAS.message(label, then));
```

Comment:
so at this point we can already assume that if there's a `has`, then there is exactly 1 constraint right? so the first check is redundant

does this pass a test you can write that should fail? i think there maybe a missing `!`, and im not sure off the top of my head if `isNamed()` is correct, or not. Now that I think about it, the `isNamed()` is for when the user writes `$abc`, and  you can check if the variable is hidden (ie not named by user) and `variable.label()` exists to check for user-written label for a variable... tho might be a better way to do that

Reply:
Should have just done .type!=null.

I think attribute.isVariable is still true though?

Reply:
i think we resolved this on the phone right?

Reply:
Yes we did :)

Reply:
`then.has().get(0).type()` should be checked with `.isPresent()` if it were an `Optional`, which it should be.